### DISCUSSION TITLE
`baseBranches` setting doesn't get migrated when using a preset

### DISCUSSION THREAD

**Discussion by DManstrator** [renovatebot/renovate#20531] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.145.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

GitLab Community Edition 15.8.1

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

In discussion #20434 I had the problem that my extended preset was ignored. I further investigated it and found out that the problem was, that the `baseBranches` setting wasn't migrated. Due to that, the preset was effectively ignored since the preset configuration only targeted a different branch.

`first_preset.json`
```json
{
  "baseBranches": [
    "special-branch"
  ],
  "packageRules": [
    [...] (only for the special-branch)
  ]
}
```

`renovate.json`
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    "local>my-group/global-config//renovate/first_preset"
  ],
  "baseBranches": [
    "develop"
  ],
  "packageRules": [
    [...]
  ],
  [...]
}
```

In the end I would have awaited, that both `baseBranches` settings would be migrated into `"baseBranches": ["develop", "special-branch"]`.

When removing the `baseBranches` section from the `renovate.json`, the preset was loaded and used.

![image](https://user-images.githubusercontent.com/7572614/220128858-8d64f94f-0e78-457b-bb75-dce8b670c714.png)

### Relevant debug logs

<details><summary>Logs (with baseBranches setting in renovate.json)</summary>

```
DEBUG: migrated config (repository=my-group/my-repo)
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "extends": [
           "config:base",
           "local>my-group/global-config//renovate/first_preset"
         ],
         "baseBranches": ["develop"],
         "enabledManagers": ["gitlabci", "maven"],
         [...]
       }
DEBUG: Setting hostRules from config (repository=my-group/my-repo)
DEBUG: Found repo ignorePaths (repository=my-group/my-repo)
       "ignorePaths": [
         "**/node_modules/**",
         "**/bower_components/**",
         "**/vendor/**",
         "**/examples/**",
         "**/__tests__/**",
         "**/test/**",
         "**/tests/**",
         "**/__fixtures__/**"
       ]
DEBUG: No vulnerability alerts found (repository=my-group/my-repo)
DEBUG: findIssue(Dependency Dashboard) (repository=my-group/my-repo)
DEBUG: baseBranches (repository=my-group/my-repo)
       "baseBranches": ["develop"]
DEBUG: baseBranch: develop (repository=my-group/my-repo, baseBranch=develop)
```
</details>

<details><summary>Logs (without baseBranches setting in renovate.json)</summary>

```
DEBUG: migrated config (repository=my-group/my-repo)
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "extends": [
           "config:base",
           "local>my-group/global-config//renovate/first_preset"
         ],
         "enabledManagers": ["gitlabci", "maven"],
         [...]
       }
DEBUG: Setting hostRules from config (repository=my-group/my-repo)
DEBUG: Found repo ignorePaths (repository=my-group/my-repo)
       "ignorePaths": [
         "**/node_modules/**",
         "**/bower_components/**",
         "**/vendor/**",
         "**/examples/**",
         "**/__tests__/**",
         "**/test/**",
         "**/tests/**",
         "**/__fixtures__/**"
       ]
DEBUG: No vulnerability alerts found (repository=my-group/my-repo)
DEBUG: findIssue(Dependency Dashboard) (repository=my-group/my-repo)
DEBUG: baseBranches (repository=my-group/my-repo)
       "baseBranches": ["special-branch"]
DEBUG: baseBranch: special-branch (repository=my-group/my-repo, baseBranch=special-branch)
```
</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
What do you mean by "migrated"?

**Comment by rarkins** [comment]
If you mean _merged_, then this is not a bug. `baseBranches` is a non-mergeable field

**Comment by DManstrator** [comment]
Yes, "merged" is what I meant, I used "migrated" since this is what I've seen in the logs. Sorry for the false bug report.

Is there any way to achieve what I want then? Imagine having a repository `repoA` with two base branches (`branchA` (default branch) and `branchB`). Is there any way to enable more base branches via. presets? As soon the `renovate.json` defines the `baseBranches` section (which would be needed in that case from what I know), then other base branches will get ignored.

In that case, I would have to add those two base branches to the preset even they don't really belong there. And if other repositories (eg. `repoB`) want to use that preset, they also may not define an own `baseBranches` section. And then `repoB` would also have the two base branches from `repoA` even they might not be in that repository. So it seems it would be good if the `baseBranches` field would actually be mergeable.

I hope you understand me, else I will try to explain it in other words.

**Reply by rarkins** [reply]
`baseBranches` is intentionally non-mergeable so that it can be overridden. I don't understand why you want the ability to define wrong/incomplete `baseBranches` within the repo's `baseBranches`. If you left out `baseBranches` from the `renovate.json` the you could have the setting in presets. Wherever you define it, it must include both

**Reply by DManstrator** [reply]
> I don't understand why you want the ability to define wrong/incomplete `baseBranches` within the repo's `baseBranches`.

I have the following situation:
- `repoA` only having the default branch as a base branch.
- `repoB` having the default branch and another branch as the base branches.
- A preset for support branches, both `repoA` and `repoB` should use it.

My current workaround is that the preset has the default branch and the support branch set as base branches and the `renovate.json` files of both repos have no `baseBranches` setting anymore. This works for `repoA` but not for `repoB` because now, the other base branch of it is ignored. And if I add that other base branch of `repoB` to the preset, I will get a warning in `repoA` because is hasn't that other base branch.

Do you have any idea for me? The main goal of the preset should be that I only have to touch that and not the `renovate.json` anymore.

**Reply by rarkins** [reply]
If a repo has default branch, you don't need `baseBranches` in either the repo or a preset.

If a repo has non-default baseBranches then you should put them in the `renovate.json`.

It's not realistic to think you can have a single shared preset and not touch `renovate.json` if the repositories have different settings.
